<div style="border-radius: 15px; box-shadow: 2px 2px 4px; border: 1px solid; background:#bdc3c7; padding: 20px"> <h7 style="color:black; margin-bottom:20px"> Денис, привет!)
 Мои комментарии и замечания далее по тексту помечены различными цветами:
    
---    
<div class="alert alert-block alert-danger">
Критическое замечание, которые следует исправить.
</div> 
    
    
---
    
<div class="alert alert-block alert-info">
Замечание\совет на будущее.
</div>
    
---
    
<div class="alert alert-block alert-success">
Когда всё сделано правильно.
</div>
    
---
    
p.s.: не удаляй мои замечания, если предстоит что то доработать в проекте.</h7>

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 

Итог ревью - ты проделал хорошую работу. Почти все запросы корректные и дают правильный результат. Но есть одно задание, которое тебе предстоит доработать, чтобы я смог принять твой проект. Если возникнут вопросы по моим комментариям, пиши здесь - обязательно отвечу. Буду ждать!
</div>

<div class="alert alert-block alert-info">
    

__Комментарий джуниора__

Привет, Михаил! Благодарю за проверку работы и комментарии. Я внес исправления в свою работу, пометил их комментариями голубого цвета.
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера v2: </b> 

Привет) Все замечания учтены. Успехов в дальнейшем обучении!
</div>

# Проект по SQL

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

Задача: проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**

1. Таблица books
Содержит данные о книгах:
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.


2. Таблица authors
Содержит данные об авторах:
* author_id — идентификатор автора;
* author — имя автора.


3. Таблица publishers
Содержит данные об издательствах:
* publisher_id — идентификатор издательства;
* publisher — название издательства;


4. Таблица ratings
Содержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.


5. Таблица reviews
Содержит данные о пользовательских обзорах:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

**Схема данных**
<img src="files/scheme_1589269096.png">



Шаги для решения задачи:
1. Посчитаем, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитаем количество обзоров и среднюю оценку;
3. Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
4. Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


Алгоритм выполнения работы:
* Опишем цели исследования;
* Исследуем таблицы — выведем первые строки;
* Сделаем по одному SQL-запросу для решения каждого задания;
* Выведем результаты каждого запроса в тетрадке;
* Опишим выводы по каждой из решённых задач.

## Изучим информацию в базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine 

In [9]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

#создадим функцию для вывода SQL запроса
def sql_result(query):
    return pd.io.sql.read_sql(query, con = engine)

In [18]:
#изучим данные в таблице books:
books = """SELECT *
           FROM books;
        """

# выводим результат
display(sql_result(books).head())
print(sql_result(books).info())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None


In [19]:
#изучим данные в таблице authors:
authors = """SELECT *
           FROM authors;
        """

display(sql_result(authors).head())
print(sql_result(authors).info())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None


In [20]:
#изучим данные в таблице publishers:
# формируем запрос
publishers = """SELECT *
           FROM publishers;
        """

# выводим результат
display(sql_result(publishers).head())
print(sql_result(publishers).info())

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None


In [23]:
#изучим данные в таблице ratings:
# формируем запрос
ratings = """SELECT *
           FROM ratings;
        """

# выводим результат
display(sql_result(ratings).head())
print(sql_result(ratings).info())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None


In [22]:
#изучим данные в таблице reviews:
# формируем запрос
reviews = """SELECT *
           FROM reviews;
        """

# выводим результат
display(sql_result(reviews).head())
print(sql_result(reviews).info())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None


### Вывод

Изучены данные, выведены первые 5 строк каждой таблицы. Набор данных соответсвует описанию.

Таблицы имеют следующий размер:
* books - 1000 строк
* authors - 636 строк
* publishers - 340 строк
* ratings - 6456 строк
* reviews - 2793 строки

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 

Здесь всё хорошо. Ты корректно выгрузил данные и посмотрел их содержимое. Не помешало бы вывести еще информацию с помощью info(). Так у нас будет информация хотя бы о размерности датасетов
</div>

<div class="alert alert-block alert-info">
    

__Комментарий джуниора__

Поправил вывод таблиц и добавил общую информацию
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера v2: </b> 

Спасибо, так гораздо информативнее
</div>

_______________________________________________________________________________________________________________________________

## Анализ данных. SQL запросы

### Посчитаем, сколько книг вышло после 1 января 2000 года;

In [24]:
count_books = """ SELECT COUNT(book_id) AS number_of_books_after_date
                FROM books
                WHERE publication_date > '2000-01-01';
                """
display(sql_result(count_books))

number_of_books_after_date
0                         819

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 
    
Всё верно. Это правильный ответ

</div>

In [25]:
count_books_all = """SELECT COUNT(book_id) AS number_of_books_after_date
                   FROM books
                """
display(sql_result(count_books_all))

number_of_books_after_date
0                        1000

После 1 января 2000 года вышло 819 книг, что является значительным числом от общего количества книг в базе данных (1000)

### Для каждой книги посчитаем количество обзоров и среднюю оценку;

In [49]:
count_reviews = """SELECT books.title AS title,
                          SUBQ1.number_of_reviews,
                          ROUND(SUBQ2.average_rating, 2) AS average_rating
                   FROM books
                   LEFT JOIN 
                        (SELECT book_id,
                                COUNT(review_id) as number_of_reviews
                         FROM reviews
                         GROUP BY book_id) AS SUBQ1
                                                 ON SUBQ1.book_id = books.book_id
                   INNER JOIN 
                        (SELECT book_id,
                                AVG(rating) as average_rating
                         FROM ratings
                         GROUP BY book_id) AS SUBQ2
                                                 ON SUBQ2.book_id = books.book_id
                   GROUP BY title,
                            SUBQ1.number_of_reviews,
                            average_rating
                   ORDER BY SUBQ1.number_of_reviews DESC;
                """
display(sql_result(count_reviews).head(15))

title  number_of_reviews  \
0     Anne Rice's The Vampire Lestat: A Graphic Novel                NaN   
1   Disney's Beauty and the Beast (A Little Golden...                NaN   
2                           Essential Tales and Poems                NaN   
3                                Leonardo's Notebooks                NaN   
4    The Cat in the Hat and Other Dr. Seuss Favorites                NaN   
5                             The Natural Way to Draw                NaN   
6                             Twilight (Twilight  #1)                7.0   
7                                     Eat  Pray  Love                6.0   
8   Harry Potter and the Chamber of Secrets (Harry...                6.0   
9   Harry Potter and the Prisoner of Azkaban (Harr...                6.0   
10                          Outlander (Outlander  #1)                6.0   
11                                      The Alchemist                6.0   
12                                     The Book Thief                6.0   
13                             The Catcher in the Rye                6.0   
14  The Curious Incident of the Dog in the Night-Time                6.0   

    average_rating  
0             3.67  
1             4.00  
2             4.00  
3             4.00  
4             5.00  
5             3.00  
6             3.66  
7             3.40  
8             4.29  
9             4.41  
10            4.13  
11            3.79  
12            4.26  
13            3.83  
14            4.08

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 
    
Почти корректно, но ты вывел только те книги, у которых есть текстовые отзывы reviews (если вывести твой запрос без limit), а ведь у книг без отзывов оценки то никуда не делись и их по задаче необходимо вывести. Доработай этот момент пожалуйста

</div>

<div class="alert alert-block alert-info">
    

__Комментарий джуниора__

Поправил в объединении таблиц, теперь есть все книги, по которым есть оценка
</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера v2: </b> 

Отлично. Теперь всё корректно)
</div>

Средняя количество отзывов топ-15 книг - 6, при этом оценки не высокие, т.е. большое количество отзывов скорее всего будет приводить к средним оценкам, даже если это бестселлер. В топ-15 книг по количеству отзывов преобладает фантастика, фэнтези и приключения, часть этих книг экранизирована, возможно этим обусловена их популярность.

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

In [55]:
count_publisher = """SELECT publishers.publisher AS publisher_name,
                  COUNT (SUBQ.book_id) AS number_of_books
           FROM (
               SELECT book_id,
                      publisher_id
               FROM books
               WHERE num_pages > 50) AS SUBQ
           INNER JOIN publishers ON publishers.publisher_id = SUBQ.publisher_id
           GROUP BY publisher_name
           ORDER BY number_of_books DESC;
        """
display(sql_result(count_publisher).head(10))

publisher_name  number_of_books
0             Penguin Books               42
1                   Vintage               31
2  Grand Central Publishing               25
3          Penguin Classics               24
4          Ballantine Books               19
5                    Bantam               19
6                   Berkley               17
7        St. Martin's Press               14
8             Berkley Books               14
9                     Delta               13

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 
    

Согласен с результатом. Действительно данное издательство является лидером по выпущенным книгам

</div>

В базе данных самое часто встречающееся издательство - "Penguin Books", на втором месте Vintage, на третьем Grand Central Publishing. Эти издательства отличаются тем, что являются одними из старейших издательств, которые обладают правами на издание книг наиболее популярных авторов 20 и 21 века. 

### Определим автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [56]:
author_ratings = """SELECT authors.author AS author_name,
                  ROUND(AVG(SUBQ.average_rating), 3) AS avg_rate    
           FROM books
           LEFT JOIN (
                   SELECT book_id,
                          COUNT(rating) AS cnt,
                          AVG(rating) AS average_rating
                   FROM ratings
                   GROUP BY book_id
                   ) AS SUBQ
                ON SUBQ.book_id = books.book_id
           LEFT JOIN authors ON authors.author_id = books.author_id
           WHERE SUBQ.cnt >= 50
           GROUP BY author_name
           ORDER BY avg_rate DESC;   
        """
display(sql_result(author_ratings).head(10))

author_name  avg_rate
0                         J.K. Rowling/Mary GrandPré     4.284
1                  Markus Zusak/Cao Xuân Việt Khương     4.264
2                                     J.R.R. Tolkien     4.258
3                                  Louisa May Alcott     4.192
4                                       Rick Riordan     4.081
5                                    William Golding     3.901
6                                      J.D. Salinger     3.826
7           Paulo Coelho/Alan R. Clarke/Özdemir İnce     3.789
8  William Shakespeare/Paul Werstine/Barbara A. M...     3.788
9                                          Dan Brown     3.755

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 
    

И здесь ты не ошибся. Ты верно отфильтровал книги и рассчитал средний рейтинг авторов (но округлять так грубо не совсем хороший ход, если взять до 3х цифр после запятой, то топ-2 и топ-3 поменяются местами)

</div>

<div class="alert alert-block alert-info">
    

__Комментарий джуниора__

Поправил округление
</div>

Наибольшая средняя оценка у LДж. Роллинг - автора книг о Гарри Поттере, затем идет Дж. Толкиен. Топ-10 по оценке авторов соотносятся с топом книг. 

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [53]:
review_count_50 = """SELECT ROUND(AVG(SUBQ2.review_cnt), 2) AS average_number_of_reviews
                   FROM (
                        SELECT username  
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating_id) > 50) AS SUBQ1
                   LEFT JOIN (
                               SELECT username,
                                      COUNT(review_id) AS review_cnt
                               FROM reviews
                               GROUP BY username) AS SUBQ2
                    ON SUBQ2.username = SUBQ1.username
                """
display(sql_result(review_count_50).head())

average_number_of_reviews
0                      24.33

In [54]:
#посомтрим общее количество обзоров
review_count_all = """SELECT ROUND(AVG(SUBQ2.review_cnt), 2) AS average_number_of_reviews
                   FROM (
                        SELECT username  
                        FROM ratings
                        GROUP BY username) AS SUBQ1
                   LEFT JOIN (
                               SELECT username,
                                      COUNT(review_id) AS review_cnt
                               FROM reviews
                               GROUP BY username) AS SUBQ2
                    ON SUBQ2.username = SUBQ1.username
                """
display(sql_result(review_count_all).head())

average_number_of_reviews
0                      17.46

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24, что выше, чем общее количество обзоров от всех пользователей(17), но составляет всего 50% от "оценочной" активности. Возможно стоит как-то стимулровать пользователй писать обзоры при оценке. Среднее количество обзоров от активных пользователей - хорошая метрика, это повысит доверие к нашему продукту, сформирует сообщество вокруг сервиса и позволит другим покупателям подобрать себе книгу.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <b>Комментарии от ревьюера: </b> 
    

И это .. правильный ответ! Ты корректно отобрал пользователей и рассчитал их показатели. Еще и сравнил с общей статистикой по БД. Молодец!


В итоге, 4\5 с первой попытки - это достойный результат)

</div>

## Общий вывод

При формулировании ценностного предложения для нового продукта, стоит ориентироваться на следующие факторы:
* в качестве книг выбирать те, которые вышли после 1 января 2000, т.к. они составляют основной объем у конкурентов;
* популярностью пользуются фантастика, фэнтези и приключения, стоит ориентироваться на экранизированые книги;
* чаще встречаются книги от старейших американских и британских издательство - "Penguin Books", Vintage, Grand Central Publishing. Эти издательства обладают правами на издание книг наиболее популярных авторов 20 и 21 века.
* наибольшая средняя оценка у авторов бестселлеров, топ-10 по оценке авторов соотносятся с топом книг. Популярность одной книги автора привлекает внимание к другим. 
* Стоит как-то стимулровать пользователй писать обзоры при оценке. Среднее количество обзоров от активных пользователей - хорошая метрика, это повысит доверие к нашему продукту, сформирует сообщество вокруг сервиса и позволит другим покупателям подобрать себе книгу.